dm/dm/master/server.go

```diff
@@ -2641,6 +2641,67 @@ func (s *Server) sharedLogic(ctx context.Context, req interface{}, respPointer i
 	return true
 }
 
+func (s *Server) checkStartValidationParams(req *pb.StartValidationRequest) (
+	map[string]map[string]config.SubTaskConfig, string, bool,
+) {
+	explicitModeOrStartTime := req.Mode != nil || req.StartTime != nil
+	if req.Mode != nil {
+		mode := req.GetModeValue()
+		if mode != config.ValidationFull && mode != config.ValidationFast {
+			msg := fmt.Sprintf("validation mode should be either `%s` or `%s`",
+				config.ValidationFull, config.ValidationFast)
+			return nil, msg, false
+		}
+	}
+
+	startTime := req.GetStartTimeValue()
+	if startTime != "" {
+		if _, err := utils.ParseStartTime(startTime); err != nil {
+			return nil, "start-time should be in the format like '2006-01-02 15:04:05' or '2006-01-02T15:04:05'", false
+		}
+	}
+
+	subTaskCfgs := s.scheduler.GetSubTaskCfgsByTaskAndSource(req.TaskName, req.Sources)
+	if len(subTaskCfgs) == 0 {
+		if req.TaskName == "" {
+			return nil, fmt.Sprintf("fail to get subtask config by sources `%v`", req.Sources), false
+		}
+		return nil, fmt.Sprintf("fail to get subtask config by task name `%s` and sources `%v`",
+			req.TaskName, req.Sources), false
+	}
+
+	var enabledValidator string
+	allEnabled, noneEnabled := true, true
+	for taskName := range subTaskCfgs {
+		for sourceID := range subTaskCfgs[taskName] {
+			if s.scheduler.ValidatorEnabled(taskName, sourceID) {
```

Comment:
the check for all subtasks is not atomic, is it safe that other RPC changed the status?

Reply:
changed to [acquire subtask latch before check](https://github.com/pingcap/tiflow/pull/5534/commits/cc708a61c22b198cfdded9843bc32b7f19d9a340)